addons/mail/static/src/models/message_sender/message_sender.js

```diff
@@ -0,0 +1,278 @@
+            // Useful to do the final replace after parsing to avoid using the
+            // same tag twice if two different mentions have the same name.
+            const mentions = [];
+            for (const partner of this.thread.composer.mentionedPartners) {
+                const placeholder = `@-mention-partner-${partner.id}`;
+                const text = `@${owl.utils.escape(partner.name)}`;
+                mentions.push({
+                    class: 'o_mail_redirect',
+                    id: partner.id,
+                    model: 'res.partner',
+                    placeholder,
+                    text,
+                });
+                body = body.replace(text, placeholder);
+            }
+            for (const channel of this.thread.composer.mentionedChannels) {
+                const placeholder = `#-mention-channel-${channel.id}`;
+                const text = `#${owl.utils.escape(channel.name)}`;
+                mentions.push({
+                    class: 'o_channel_redirect',
+                    id: channel.id,
+                    model: 'mail.channel',
+                    placeholder,
+                    text,
+                });
+                body = body.replace(text, placeholder);
+            }
+            const baseHREF = this.env.session.url('/web');
+            for (const mention of mentions) {
+                const href = `href='${baseHREF}#model=${mention.model}&id=${mention.id}'`;
+                const attClass = `class='${mention.class}'`;
+                const dataOeId = `data-oe-id='${mention.id}'`;
+                const dataOeModel = `data-oe-model='${mention.model}'`;
+                const target = `target='_blank'`;
+                const link = `<a ${href} ${attClass} ${dataOeId} ${dataOeModel} ${target}>${mention.text}</a>`;
+                body = body.replace(mention.placeholder, link);
+            }
+            return body;
+        }
+
+        /**
+         * @private
+         * @param {string} htmlString
+         * @returns {string}
+         */
+        _generateEmojisOnHtml(htmlString) {
+            for (const emoji of emojis) {
+                for (const source of emoji.sources) {
+                    const escapedSource = String(source).replace(
+                        /([.*+?=^!:${}()|[\]/\\])/g,
+                        '\\$1');
+                    const regexp = new RegExp(
+                        '(\\s|^)(' + escapedSource + ')(?=\\s|$)',
+                        'g');
+                    htmlString = htmlString.replace(regexp, '$1' + emoji.unicode);
+                }
+            }
+            return htmlString;
+        }
+
+        async _postMailChannel(postData) {
+            let messageId;
+            const command = this.thread.composer._getCommandFromText(postData.body);
+            Object.assign(postData, {
+                subtype_xmlid: 'mail.mt_comment',
+            });
+            if (command) {
```

Comment:
I think command is too far in the chain.

I don't even know if we want to actually queue commands at all.
For me a channel command is just a shortcut that replaces a button click.
It's not a message, and should probably never be a pending message either.

Unless we somehow want to start queuing every single possible rpc?

Reply:
> Unless we somehow want to start queuing every single possible rpc?

We do ?
If you send a message, then you send a /quit, do you want to have the possibility for the /quit to be executed before the message ? It's very unlikely to happen, I know, but it can if the command RPC is for some raison, faster than the message RPC.

Reply:
> What is to say when clicking on a button and then another one, the order isn't also important?

It can be if the button are doing thing that are somehow, related. But it could also be a bad UI that use too much button.

> Should we completely block the UI after any operation? Or actually queue all RPC at a lower level instead?

Queuing every RPC could be a nice feature but I don't know if it's useful for many use-case.

Reply:
At first glance it would look detrimental to most "get" rpc but would make sense for most "post" rpc.
(I'm talking in terms of functional "get" or "post" because in practice our jsonrpc implementation is always POST).

Reasoning is that "get" should normally be without side effect and perfectly fine to run in any order.
And with "post" the user is doing interactions in a specific order, so it should ideally happen in that order too.
But even if the queue we still have the problem that the result of a first interaction can change the state of the application in a way that would make following interactions potentially obsolete.

That being said I wonder if we shouldn't immediately make the queue independent of what kind of data it is sending, even just for the sake of proper split of concern? The current `_processMessagesToBeSent` is (too?) long and complex because it handles both the rpc queue and the message logic at the same time.